text/0038-scrubbing-sensitive-data.md

```diff
@@ -0,0 +1,135 @@
+
+# Summary
+
+Make scrubbing of sensitive data (security related data (passwords, keys, et al) and personal identifiable information (PII)) in Sentry smarter.
+This includes the SDK side (before we send data) and the Relay side (before we ingest data)
+
+# Motivation
+
+Currently the scrubbing of sensitive data works, but is not very smart. It can happen that sensitive data is leaked and ends up in our data storage.
+It also can happen, that scrubbing is too broadly defined and too much data is removed so the feature distroys value for the customer.
+
+We want to have a smarter way of scrubbing sensitive data with a more fine grained control of what data is preserved and what data is scrubbed.
+
+We do this to make sure we handle our users sensitive data with the respect it deserves.
+
+# Background
+
+A user complained that our data scrubbing did not remove sensitive data from an URL the users project made a HTTP request to. The sensitive data showed up in the span description in the Sentry UI.
+
+# Supporting Data
+
+(tbd)
+
+# Options Considered
+
+### Status Quo:
+
+- Right now most data scrubbing is done in Relay.
+- There is a option `sendDefaultPII` in SDKs that may or may not remove some sensitive data before sending.
+- In Relay fields have a "pii" attribute that can be `yes`, `no` or `maybe`:
+  - `yes` means all the default data scrubbing Regexes are applied to the fields.
+  - `maybe` means that the data scrubbing Regexes are not run by default. If the user has advanced data scrubbing set (in Sentry.io under: Project Settings > Security & Privacy > Advanced Data Scrubbing) those custom rules are applied to the field and if a rule matches the whole content of the field is removed.
+  - `no` there is no possibility of data scrubbing for this field.
+- The regexes for data scrubbing are defined [here](https://github.com/getsentry/relay/blob/master/relay-general/src/pii/regexes.rs#L103-L274)
+- Some regexes can just remove the sensitive part of the content (like IP and SSH keys regexes).
+- Some of the regexes (like password regex) will remove the complete content of a field. This is because it is unstructured data. Relay does not know if the content is a SQL query, a JSON object, an URL, a Elasticsearch/MongoDB/whatever query in JSON format, or something else.
+
+### Currently perferred solution:
+
+A combination of some of the options listed below:
+
+- Have the SDKs store structured data (Option B)
+- If Relay does not get any structured data from Option B, Relay does Option C.
+- If Relay can not figure out what kind of data is in the field, do matching like we do now.
+
+### Option A): Remove Sensitive Data in SDKs
+
+At the time when SDKs set the data in an event, it knows what the data represents and can remove sensitive information. This way Relay has to beliefe that the SDK does the right thing and does not need to scrub data.
+
+_Pros:_
+
+-
+
+_Cons:_
+
+-
+
+### Option B): Store data in a structured way in SDKs
```

Comment:
The other point is that it'd only work for auto instrumentation (because we apply the structured data) or when people follow our guidelines, but when they don't, there might be PII, what do we do in such cases?